### ISSUE TITLE
Incorrect "mismatched types [E0308]" error

### ISSUE THREAD

**Issue by AlyoshaVasilieva** [intellij-rust/intellij-rust#4613]
## Environment

* **IntelliJ Rust plugin version:** 0.2.110.5731-193-dev
* **Rust toolchain version:** 1.39 stable
* **IDE name and version:** IntelliJ IDEA 2019.3 beta
* **Operating system:** Windows 10 x64

## Problem description
Plugin gives incorrect E0308. New experimental macro engine is enabled.

## Steps to reproduce

```rust
fn main() {
    eat_slice(&[]); // error highlights empty slice ref
}

fn eat_slice(slice: &[&str]) {}
```

Error is:
```
mismatched types [E0308]
expected `&[&str]`, found `&[<unknown>]`
```

Code compiles without error.

**Comment by alexschrod** [comment]
I don't know if I have any experimental macro engine enabled, but ever since an update or two ago, I've also been having this issue.
![image](https://user-images.githubusercontent.com/767490/68534193-a97c5500-0331-11ea-807f-5de0f51cce4b.png)
In this case, `Offscreen` implements `SafeConsole`, and this code used to not show errors. Compiles and runs just fine.

**Comment by serid** [comment]
Same here. Code compiles and runs and `String` is, indeed a dereferenced version of `&str`.

Maybe the plugin could use rls (rust language server)?
https://github.com/rust-lang/rls

![image](https://user-images.githubusercontent.com/30107608/68595249-20d4f480-04aa-11ea-867c-477264f4eb2c.png)

**Comment by Undin** [comment]
> Same here. Code compiles and runs and String is, indeed a dereferenced version of &str.

@serid It's not the same. See #4224 for more details and how to fix it

> Maybe the plugin could use rls (rust language server)?

No, we aren't going to use rls

**Comment by vlad20012** [comment]
@AlyoshaVasilieva Thanks for the report! I fixed it in #4621.

@alexschrod please, link the project you are working on or reduce it to a minimal example. And your issue is different

**Comment by alexschrod** [comment]
@vlad20012 I dunno, whatever you did to fix this issue also appears to have fixed mine. 🤷‍♂ Or at the very least, the issue I had is gone again with the latest release.
